### PR TITLE
MAINT: Deprecate NumPy functions in SciPy root

### PR DESCRIPTION
**Pull request by larsoner**
From discussion in #10238, it came up that we should probably deprecate support for `scipy.<NumPy function>`. This PR utilizes ~~sklearn's `deprecated` [decorator](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/deprecation.py)~~ a minimal `deprecated` decorator to deprecate them, marking them for removal in ~~1.5.0~~ 2.0, with `scipy.fft` removed in 1.5.0.

### CODE REVIEW

File: doc/release/1.4.0-notes.rst [hunk 1]
```diff
@@ -76,6 +76,13 @@ Deprecated features
 Backwards incompatible changes
 ==============================
 
+`scipy` changes
+---------------
+Support for NumPy functions exposed via the root SciPy namespace is deprecated
+and will be removed in 1.5.0. For example, if you use ``scipy.fft`` or
+``scipy.diag``, you should change your code to directly use :func:`numpy.fft`
```
Comment (peterbell10) [inline doc/release/1.4.0-notes.rst]:
```suggestion
``scipy.diag``, you should change your code to directly use :func:`numpy.fft.fft`
```

**Review by peterbell10 — COMMENTED** [review]
If this is deprecated, will it still need to be supported in code that imports `scipy.fft`? Considering that scipy.fft is only going to be imported be new code.

**Comment by larsoner** [comment]
@peterbell10 I don't quite understand your question.

For what it's worth, after this PR, all you should need to add to #10238 to get `scipy.ftt`-as-callable deprecated is to swap in `deprecated(...)(np.fft)` whereever you currently just use `np.fft`.

**Comment by peterbell10** [comment]
To illustrate the point:
```python
>>>import scipy
>>>scipy.fft
<function fft ...>
>>>import scipy.fft
<module 'scipy.fft' ...>
```

So, if the code doesn't import `scipy.fft` this should never come up as an issue.

File: doc/release/1.4.0-notes.rst [hunk 2]
```diff
@@ -76,6 +76,13 @@ Deprecated features
 Backwards incompatible changes
 ==============================
 
+`scipy` changes
+---------------
+Support for NumPy functions exposed via the root SciPy namespace is deprecated
```
Comment (WarrenWeckesser) [inline doc/release/1.4.0-notes.rst]: This should be in a "Deprecated features" section.  When the names are actually removed, it is noted in that release under the "Backwards incompatiable changes" section.

**Comment by larsoner** [comment]
I still don't understand. What is the problematic code snippet / behavior (once this PR is merged) that wouldn't be solved by swapping in the `deprecated(...)` for `np.fft` in your module-as-callable in #10238?

**Comment by ilayn** [comment]
The problem is when someone writes `from scipy import fft` which one is going to be imported. 

I feel like we are going to get quite some complaint about this but I guess better now than later.

**Comment by WarrenWeckesser** [comment]
Avoid the potential confusion with `fft` by using a different example, e.g. `scipy.array` -> `numpy.array`.

**Comment by peterbell10** [comment]
Ah, okay. I hadn't considered the ambiguity with `from scipy import fft`.

**Comment by larsoner** [comment]
> The problem is when someone writes from scipy import fft which one is going to be imported.

Ahh I see what you mean. But this is not really an issue with this PR but rather #10238.

**Comment by rgommers** [comment]
>

**Comment by rgommers** [comment]
> marking them for removal in 1.5.0.

This can't be right, the minimum is 2 releases worth of deprecation warnings and this PR is for 1.4.0. This is important enough to discuss on the mailing list. I think it's a cleanup worth doing, but I'm doubting between 1.6.0 and 2.0.0 for removal.

**Comment by larsoner** [comment]
Bumped to 1.6.0 and mailing list message sent

**Comment by WarrenWeckesser** [comment]
The use of the numpy names from `scipy` might be more pervasive than we expect.  For example, quite a few of the older Cookbook notebooks use them:

* https://scipy-cookbook.readthedocs.io/items/Matplotlib_converting_a_matrix_to_a_raster_image.html
* https://scipy-cookbook.readthedocs.io/items/Interpolation.html
* https://scipy-cookbook.readthedocs.io/items/CommTheory.html
* https://scipy-cookbook.readthedocs.io/items/LinearRegression.html
* https://scipy-cookbook.readthedocs.io/items/MayaVi_mlab.html
* https://scipy-cookbook.readthedocs.io/items/MayaVi_ScriptingMayavi2.html
* https://scipy-cookbook.readthedocs.io/items/OptimizationAndFitDemo1.html
* https://scipy-cookbook.readthedocs.io/items/SignalSmooth.html
* https://scipy-cookbook.readthedocs.io/items/SphericalBesselZeros.html

Examples show up on stackoverflow now and then (some based on those cookbook notebooks, and they might be mostly older questions):

* (2017) https://stackoverflow.com/questions/43317119/python-importerror-numpy-mkl (note the line `from scipy import *` in the error message)
* (2013) https://stackoverflow.com/questions/20612769/how-to-get-sum-of-each-row-and-sum-of-each-column-in-scipy-sparse-matrices-csr
* (2013) https://stackoverflow.com/questions/19849167/find-most-distant-points-in-contour-curve/ (see the answer)
* (2012) https://stackoverflow.com/questions/11401281/n-d-interpolation-for-equally-spaced-data
* (2009) https://stackoverflow.com/questions/1261578/substituting-in-a-file

(Don't let the small sample bias your judgement of the prevalence of the use the numpy names in `scipy`.  I found those with just a little bit of searching; it was definitely not an exhaustive search.  I have no idea how common it really is.)

I'm not drawing a conclusion at the moment, just sharing the result of a little bit of searching.

**Comment by rgommers** [comment]
> The use of the numpy names from scipy might be more pervasive than we expect.

My expectation is actually that there's a lot of code out there, hence my suggestion 2.0 may be better than 1.6. Also, people don't read deprecation warnings anyway .....

Another data point is the many people that are currently opening issues for the `misc` functions we removed in 1.3.0 after those having been deprecated since 0.19.0 or 1.0.0. 

> mailing list message sent

@larsoner either it's stuck in your outbox, or it got swallowed on the way somewhere. not in the archives or my inbox after 4 hours

**Comment by larsoner** [comment]
Whoops, sent it to the old `scipy.org` email! Now you should see it:

https://mail.python.org/pipermail/scipy-dev/2019-June/023551.html

**Comment by larsoner** [comment]
Also, I can live with 2.0. I'm still hoping that these use cases are not so prevalent nowadays so we can stick with 1.6.0, but if not that's okay.

**Comment by peterbell10** [comment]
Just a thought: could `scipy.fft == numpy.fft.fft` be removed in an earlier version than the others? It's clearly the most problematic and it also isn't explicitly documented anywhere I could find.

**Comment by rgommers** [comment]
> Just a thought: could scipy.fft == numpy.fft.fft be removed in an earlier version than the others? It's clearly the most problematic and it also isn't explicitly documented anywhere I could find.

I think that's fine, given that making the chance will give a clear and easy to fix exception.

File: scipy/__init__.py [hunk 3]
```diff
@@ -65,13 +65,37 @@
...
-# Import numpy symbols to scipy name space
+# Import numpy symbols to scipy name space (DEPRECATED)
+from ._lib.deprecation import _deprecated
 import numpy as _num
 linalg = None
-from numpy import *
+_msg = ('scipy.{0} is deprecated and will be removed in SciPy 2.0.0, '
+        'use numpy.{0} instead')
+for _key in dir(_num):
+    if not _key.startswith('_'):
```
Comment (pv) [inline scipy/__init__.py]: `for _key in _num.__all__:`

File: scipy/__init__.py [hunk 4]
```diff
@@ -65,13 +65,37 @@
...
-from numpy.lib.scimath import *
+_msg = ('scipy.{0} is deprecated and will be removed in SciPy 2.0.0, '
+        'use numpy.fft.{0} instead')
+fft = _deprecated(_msg.format('fft'))(fft)
+ifft = _deprecated(_msg.format('ifft'))(ifft)
+import numpy.lib.scimath as _sci
+_msg = ('scipy.{0} is deprecated and will be removed in SciPy 2.0.0, '
+        'use numpy.lib.scimath.{0} instead')
+for _key in dir(_sci):
+    if not _key.startswith('_'):
```
Comment (pv) [inline scipy/__init__.py]: `for _key in _sci.__all__:`

File: scipy/_lib/deprecation.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import warnings
+
+__all__ = ["_deprecated"]
+
+
+def _deprecated(msg):
+
+    def wrap(fun):
+        def call(*args, **kwargs):
```
Comment (pv) [inline scipy/_lib/deprecation.py]:
```
@functools.wraps(fun)
def call(*args, **kwargs):
    ...
```

File: doc/release/1.4.0-notes.rst [hunk 6]
```diff
@@ -73,6 +73,13 @@ lambda expressions for them.
 Deprecated features
 ===================
 
+`scipy` deprecations
+--------------------
+Support for NumPy functions exposed via the root SciPy namespace is deprecated
+and will be removed in 2.0.0. For example, if you use ``scipy.rand`` or
```
Comment (pv) [inline doc/release/1.4.0-notes.rst]: Maybe add a reassuring sentence here:
`... 2.0.0. They remain available in the currently continuing Scipy 1.x release series.`

Should also be added here:
https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.4.0

**Review by pv — COMMENTED** [review]
The 2.0 release idea is probably OK.
Some minor comments though.

**Comment by larsoner** [comment]
@pv comments addressed, including some clarifications in the release notes and copy-pasting them to the wiki.

@rgommers @peterbell10 I made it so that `scipy.fft` is deprecated and will be removed in 1.5.0 so we don't have to live with the module-as-callable hack for as long.

**Comment by larsoner** [comment]
Removing `needs-decision` label since it sounds like people are happy with the 2.0 in general (with 1.5 just for `scipy.fft`) removal, but if others disagree feel free to chime in.

File: scipy/_lib/deprecation.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+__all__ = ["_deprecated"]
+
+
+def _deprecated(msg):
+    """Deprecate a function by emitting a warning on use."""
+    def wrap(fun):
+        @functools.wraps(fun)
+        def call(*args, **kwargs):
+            warnings.warn(msg, category=DeprecationWarning)
```
Comment (pv) [inline scipy/_lib/deprecation.py]: This probably needs `stacklevel=2`?

Comment (larsoner) [reply scipy/_lib/deprecation.py]: You were correct with `stacklevel=2` originally, test added

**Comment by larsoner** [comment]
Actually I suspect those will be fixed by #10482, I'll merge that and rebase this

**Comment by larsoner** [comment]
All better!

**Comment by larsoner** [comment]
([OSX failure on Travis](https://travis-ci.org/scipy/scipy/jobs/560674038) looks spurious so I'll restart)

**Comment by rgommers** [comment]
@larsoner needs another rebase

**Comment by larsoner** [comment]
Okay, rebased

**Comment by rgommers** [comment]
Did some testing, all works as expected. In it goes, thanks @larsoner!
